apps/manager_app/user_training.c

```diff
@@ -0,0 +1,285 @@
+                .user_action = JS_ACTION_RIGHT,
+                .status = MANAGER_USER_TRAINING_RIGHT},
+               {.instruction = "Toggle down your joystick",
+                .user_action = JS_ACTION_DOWN,
+                .status = MANAGER_USER_TRAINING_DOWN},
+               {.instruction = "Toggle left your joystick",
+                .user_action = JS_ACTION_LEFT,
+                .status = MANAGER_USER_TRAINING_LEFT},
+               {.instruction = "Center click your joystick",
+                .user_action = JS_ACTION_CENTER,
+                .status = MANAGER_USER_TRAINING_CENTER}}};
+  evt_config_t *e_config = &context.evt_config;
+  context.events = get_events(e_config->evt_selection, e_config->timeout);
+
+  return context;
+}
+
+static bool validate_training_request(training_context_t *ctx) {
+  if (true == ctx->events.p0_event.flag) {
+    // timeout cannot happen during flow entry, or host aborted. Both are
+    // unexpected p0 event, do not proceed in flow
+    if (!ctx->events.p0_event.abort_evt) {
+      SEND_ERROR(&ctx->training_resp, 1);    // TODO: use correct error code
+    } else {
+      usb_clear_event();
+    }
+    return false;
+  }
+
+  // verify host query data is correct
+  manager_train_user_request_t *req = &ctx->training_req->train_user;
+  if (MANAGER_QUERY_TRAIN_USER_TAG != ctx->training_req->which_request ||
+      MANAGER_TRAIN_USER_REQUEST_INITIATE_TAG != req->which_request) {
+    SEND_ERROR(&ctx->training_resp, 1);    // TODO: use correct error code
+    return false;
+  }
+
+  // TODO: check the request with what device knows
+  return true;
+}
+
+static void fill_training_error(manager_train_user_response_t *training,
+                                uint32_t error_code) {
+  training->which_response = MANAGER_TRAIN_USER_RESPONSE_CORE_ERROR_TAG;
+  training->core_error.which_error = ERROR_CORE_ERROR_DEVICE_SETUP_REQUIRED_TAG;
+  training->core_error.device_setup_required = error_code;
+}
+
+static void send_msg_to_host(manager_train_user_response_t *training) {
+  uint8_t payload[MANAGER_GET_DEVICE_INFO_RESULT_RESPONSE_SIZE] = {0};
+  manager_result_t msg = MANAGER_RESULT_INIT_ZERO;
+  size_t msg_size = 0;
+  msg.which_response = MANAGER_RESULT_TRAIN_USER_TAG;
+
+  memcpy(&msg.train_user, training, sizeof(manager_train_user_response_t));
+  encode_manager_result(
+      &msg, payload, MANAGER_GET_DEVICE_INFO_RESULT_RESPONSE_SIZE, &msg_size);
+  usb_send_msg(payload, msg_size);
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+void manager_user_training(manager_query_t *query) {
+  training_context_t ctx = init_context(query);
+  core_status_set_idle_state(CORE_DEVICE_IDLE_STATE_USB);
+  core_status_set_device_waiting_on(CORE_DEVICE_WAITING_ON_IDLE);
+
+  if (false == validate_training_request(&ctx)) {
```

Comment:
Hey Ujjwal, as we discussed earlier.. can you remove this function call from here?

In case a P0 event occurred in the background.. get events will return without changing anything.. and this corner case will be handled

Reply:
Okay, it's understandable. will remove the redundant checks

Reply:
removed the check